sandbox/cgroup/kill.go

```diff
@@ -0,0 +1,122 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package cgroup
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"syscall"
+)
+
+// KillSnapProcesses sends a signal to all the processes belonging to
+// a given snap.
+//
+// A correct implementation is picked depending on cgroup v1 or v2 use in the
+// system. When cgroup v1 is detected, the call will directly act on the freezer
+// group created when a snap process was started, while with v2 the call will
+// act on all tracking groups of a snap.
+//
+// XXX: It is important to note that the implementation for v2 is racy because
+// processes can be killed externally even when the cgroup is frozen and have
+// their pid reused. Also, v2 freezing support is a no-op on kernels before
+// 5.2 which means new processes can keep getting spawned while killing
+// pids read earlier.
+var KillSnapProcesses = killSnapProcessesImplV1
```

Comment:
How is this selected at runtime?

Separately, I think we should default to v2 and only fallback to v1 if an expected interface file is missing.

Reply:
It is picked in the init for the cgroup package
https://github.com/canonical/snapd/blob/df2ffc71505a3a7a0566bb0c61bf5c81c0c0177b/sandbox/cgroup/cgroup.go#L64-L72
https://github.com/canonical/snapd/pull/14354/files/089e9df70b9dea0ecd9c74f7033d57bb10283713#diff-aab845d18e1a63cdd61215ac4596638ea81e894078faaff2a75c7f6cd52641aeR77

Reply:
Perhaps it should be initialized to a dummy function that panics or to nil. Not sure if this is worth it but it might help with casual readability that it's not the final function.